ledger/src/blockstore_processor.rs

```diff
@@ -262,40 +262,100 @@ fn execute_batches_internal(
     entry_callback: Option<&ProcessCallback>,
     transaction_status_sender: Option<&TransactionStatusSender>,
     replay_vote_sender: Option<&ReplayVoteSender>,
-    timings: &mut ExecuteTimings,
+    cumulative_timings: &mut ExecuteTimings,
+    end_to_end_timings: &mut ThreadExecuteTimings,
     cost_capacity_meter: Arc<RwLock<BlockCostCapacityMeter>>,
     tx_costs: &[u64],
 ) -> Result<()> {
     inc_new_counter_debug!("bank-par_execute_entries-count", batches.len());
-    let (results, new_timings): (Vec<Result<()>>, Vec<ExecuteTimings>) =
-        PAR_THREAD_POOL.install(|| {
-            batches
-                .into_par_iter()
-                .enumerate()
-                .map(|(index, batch)| {
-                    let mut timings = ExecuteTimings::default();
-                    let result = execute_batch(
-                        batch,
-                        bank,
-                        transaction_status_sender,
-                        replay_vote_sender,
-                        &mut timings,
-                        cost_capacity_meter.clone(),
-                        tx_costs[index],
-                    );
-                    if let Some(entry_callback) = entry_callback {
-                        entry_callback(bank);
-                    }
-                    (result, timings)
-                })
-                .unzip()
-        });
-    timings.saturating_add_in_place(ExecuteTimingType::TotalBatchesLen, batches.len() as u64);
-    timings.saturating_add_in_place(ExecuteTimingType::NumExecuteBatches, 1);
-    for timing in new_timings {
-        timings.accumulate(&timing);
+    let execution_timings_per_thread: RwLock<HashMap<usize, ThreadExecuteTimings>> =
+        RwLock::new(HashMap::new());
+
+    let results: Vec<Result<()>> = PAR_THREAD_POOL.install(|| {
+        batches
+            .into_par_iter()
+            .enumerate()
+            .map(|(index, batch)| {
+                let transaction_count = batch.sanitized_transactions().len() as u64;
+                let mut timings = ExecuteTimings::default();
+                let (result, execute_batches_time) = Measure::this(
+                    |_| {
+                        let result = execute_batch(
+                            batch,
+                            bank,
+                            transaction_status_sender,
+                            replay_vote_sender,
+                            &mut timings,
+                            cost_capacity_meter.clone(),
+                            tx_costs[index],
+                        );
+                        if let Some(entry_callback) = entry_callback {
+                            entry_callback(bank);
+                        }
+                        result
+                    },
+                    (),
+                    "execute_batch",
+                );
+
+                let thread_index = PAR_THREAD_POOL.current_thread_index().unwrap();
```

Comment:
How will you know if it's rayon just scheduled a bunch of small jobs on one thread vs. a single (or multiple) large batch(es) which was executed on one thread. Also temporally if that created the longest timings or maybe that thread started sooner than another thread and actually caused the batch to take longer. And the rayon scheduling could vary across machines, no?

Reply:
>  if it's rayon just scheduled a bunch of small jobs on one thread vs. a single (or multiple) large batch(es) which was executed on one thread

Yeah we won't be able to distinguish this, we'll just take the longest running thread and hopes that captures the bottleneck

>  Also temporally if that created the longest timings or maybe that thread started sooner than another thread and actually caused the batch to take longer. And the rayon scheduling could vary across machines, no?

I was thinking if the longest thread's total thread execution time doesn't add up to the total replay time, then we know there's some rayon shenanigans going on.

Reply:
Could we stick beginning and end ticks in here to gain insight into any possible Rayon scheduling shenanigans?

Reply:
I'm wondering if recording either start or end time for threads and snapshotting for the longest running thread could help understand cases where longest thread time is shorter than total execution time. Could shed some light on if that thread got started late, for example